### PR TITLE
Don't 404 `deep-reblame` on renamed files

### LINKED ISSUE #2949: `deep-blame` doesn't support renamed files
Feature: https://github.com/sindresorhus/refined-github/pull/2881

1. Visit https://github.com/sindresorhus/refined-github/blame/fad67fb8d8fcf608601c9b8ae1b5f12337cbccce/source/options.html#L1
2. Click Deep Blame button
3. See 404 page

The file is currently `source/options.html` but was [renamed](https://github.com/sindresorhus/refined-github/commits/fad67fb8d8fcf608601c9b8ae1b5f12337cbccce/src/options.html) [twice](https://github.com/sindresorhus/refined-github/commits/fad67fb8d8fcf608601c9b8ae1b5f12337cbccce/extension/options.html).

The link we generate should point to the file that existed at the time: `extension/options/index.html`

The regular blame follows renames, but I'm not sure we (`deep-blame`) can.

### PR DESCRIPTION
**Pull request by yakov116**
Fixes #2949

@fregante how do you want to handle the message?

# Test

Related PR in commit title: it should "enter" it https://github.com/sindresorhus/refined-github/blame/b5497da33a6c5b99eb518ecc266881780bc11877/contributing.md#L7

Unrelated PR in commit title: it should `alert()` https://github.com/sindresorhus/refined-github/blame/d1e27cb34f18b9d5d57fdbf853e5b15c3a660d85/source/features/deep-reblame.tsx#L67

Renamed file: it should `alert()` 
https://github.com/sindresorhus/refined-github/blame/fad67fb8d8fcf608601c9b8ae1b5f12337cbccce/source/options.html#L1
<!--

Thanks for contributing! 🍄

1. LINKED ISSUES:
   Does this PR close/fix an existing issue? Write something like `Closes #10`

2. TEST URLS:
   Add some test URLs

3. SCREENSHOT:
   Add a screenshot here if your PR makes visual changes

### CODE REVIEW

File: source/features/deep-reblame.tsx [hunk 1]
```diff
@@ -68,7 +73,9 @@ async function redirectToBlameCommit(event: DelegateEvent<MouseEvent, HTMLAnchor
 
 	blameElement.blur(); // Hide tooltip after click, it’s shown on :focus
 
-	const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber);
+	const currentFilename = getCleanPathname().split('/').slice(-3).join('/');
```
Comment (fregante) [inline source/features/deep-reblame.tsx:76]: This fails on branches with slashes. Try with `getReference` (as used in other features)

Maybe just:
```suggestion
	const [, currentFilename] = getCleanPathname().split(getReference());
```

File: source/features/deep-reblame.tsx [hunk 2]
```diff
@@ -36,7 +41,7 @@ const getPullRequestBlameCommit = mem(async (commit: string, prNumber: number):
 
 	const associatedPR = repository.object.associatedPullRequests.nodes[0];
 
-	if (!associatedPR) {
+	if (!associatedPR || !repository.file) {
 		return false;
 	}
```
Comment (fregante) [inline source/features/deep-reblame.tsx:46]: I think you'll have to throw instead so in `redirectToBlameCommit` you can output a specific message.
```suggestion
	if (!associatedPR) {
		throw new Error('The PR linked in the title didn’t create this commit');
	}
	
	if (!repository.file) {
		throw new Error('The file was renamed and Refined GitHub can’t find it');
	}
```

File: source/features/deep-reblame.tsx [hunk 3]
```diff
@@ -68,7 +73,9 @@ async function redirectToBlameCommit(event: DelegateEvent<MouseEvent, HTMLAnchor
 
 	blameElement.blur(); // Hide tooltip after click, it’s shown on :focus
 
-	const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber);
+	const [, currentFilename] = getCleanPathname().split(getReference());
+
+	const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber, currentFilename);
 	if (prBlameCommit) {
```
Comment (fregante) [inline source/features/deep-reblame.tsx:79]:
```suggestion
	try {
		const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber, currentFilename);
```

**Comment by fregante** [comment]
Follow the PR template...

**Comment by yakov116** [comment]
> Follow the PR template...

Which part did I miss? The link?

File: source/features/deep-reblame.tsx [hunk 4]
```diff
@@ -37,7 +42,11 @@ const getPullRequestBlameCommit = mem(async (commit: string, prNumber: number):
 	const associatedPR = repository.object.associatedPullRequests.nodes[0];
 
 	if (!associatedPR) {
```
Comment (fregante) [inline source/features/deep-reblame.tsx:44]:
```suggestion
	if (associatedPR?.number !== prNumber || associatedPR.mergeCommit.oid !== commit) {
```

This way you can avoid the duplicate identical error

File: source/features/deep-reblame.tsx [hunk 5]
```diff
@@ -68,25 +77,26 @@ async function redirectToBlameCommit(event: DelegateEvent<MouseEvent, HTMLAnchor
...
 
-	const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber);
-	if (prBlameCommit) {
+	const [, currentFilename] = getCleanPathname().split(getReference()! + '/');
+
+	try {
+		const prBlameCommit = await getPullRequestBlameCommit(prCommit, prNumber, currentFilename);
 		const lineNumber = select('.js-line-number', blameHunk)!.textContent!;
-		const href = new URL(location.href.replace(getReference()!, prBlameCommit));
+		const href = new URL(location.href.replace(getReference()!, prBlameCommit as string));
```
Comment (fregante) [inline source/features/deep-reblame.tsx:79]:
```suggestion
		const href = new URL(location.href.replace(getReference()!, prBlameCommit));
```

File: source/features/deep-reblame.tsx [hunk 6]
```diff
@@ -7,11 +7,16 @@ import delegate, {DelegateEvent} from 'delegate-it';
 import * as api from '../libs/api';
 import features from '../libs/features';
 import loadingIcon from '../libs/icon-loading';
-import {getRepoGQL, getReference, looseParseInt} from '../libs/utils';
+import {getRepoGQL, getReference, looseParseInt, getCleanPathname} from '../libs/utils';
 
-const getPullRequestBlameCommit = mem(async (commit: string, prNumber: number): Promise<string | false> => {
+const getPullRequestBlameCommit = mem(async (commit: string, prNumber: number, currentFilename: string): Promise<string | Error> => {
```
Comment (fregante) [inline source/features/deep-reblame.tsx:12]: This only returns `string`; errors aren't typed 

```suggestion
const getPullRequestBlameCommit = mem(async (commit: string, prNumber: number, currentFilename: string): Promise<string> => {
```

File: source/features/deep-reblame.tsx [hunk 7]
```diff
@@ -68,25 +77,26 @@ async function redirectToBlameCommit(event: DelegateEvent<MouseEvent, HTMLAnchor
...
-	}
+			blameElement.setAttribute('aria-label', 'View blame prior to this change.');
+			blameElement.classList.remove('rgh-deep-blame');
+			spinner.replaceWith(versionIcon());
+		} else {
+			spinner.remove();
+		}
 
-	alert('The PR linked in the title didn’t create this commit');
+		alert(error);
```
Comment (fregante) [inline source/features/deep-reblame.tsx:92]:
```suggestion
		alert(error.message);
```

File: source/features/deep-reblame.tsx (hunk 4 above)
Comment (yakov116) [reply source/features/deep-reblame.tsx]: Done

File: source/features/deep-reblame.tsx [hunk 8]
```diff
@@ -41,7 +41,7 @@ const getPullRequestBlameCommit = mem(async (commit: string, prNumber: number, c
 
 	const associatedPR = repository.object.associatedPullRequests.nodes[0];
 
-	if (associatedPR?.number !== prNumber || associatedPR.mergeCommit.oid !== commit) {
+	if (!associatedPR || associatedPR.number !== prNumber || associatedPR.mergeCommit.oid !== commit) {
```
Comment (yakov116) [inline source/features/deep-reblame.tsx:44]: I did that ti not have to do ? by all of them

Comment (fregante) [reply source/features/deep-reblame.tsx:44]: Fair, it is more clear 👍

**Comment by fregante** [comment]
# 🎉

File: source/features/deep-reblame.tsx [hunk 9]
```diff
@@ -54,14 +54,12 @@ const getPullRequestBlameCommit = mem(async (commit: string, prNumber: number, c
 
 async function redirectToBlameCommit(event: DelegateEvent<MouseEvent, HTMLAnchorElement | HTMLButtonElement>): Promise<void> {
 	const blameElement = event.delegateTarget;
-	if (blameElement instanceof HTMLAnchorElement) {
-		if (event.altKey) {
-			event.preventDefault();
-		} else {
-			return; // Unmodified click on regular link: let it proceed
-		}
+	if (blameElement instanceof HTMLAnchorElement && !event.altKey) {
```
Comment (yakov116) [inline source/features/deep-reblame.tsx:57]: I had it like that
https://github.com/sindresorhus/refined-github/pull/2881#pullrequestreview-378184239

File: source/features/deep-reblame.tsx [hunk 10]
```diff
@@ -68,25 +69,26 @@ async function redirectToBlameCommit(event: DelegateEvent<MouseEvent, HTMLAnchor
...
-		spinner.replaceWith(versionIcon());
-	} else {
-		spinner.remove();
-	}
+			blameElement.setAttribute('aria-label', 'View blame prior to this change.');
+			blameElement.classList.remove('rgh-deep-blame');
+			spinner.replaceWith(versionIcon());
+		} else {
+			spinner.remove();
+		}
```
Comment (fregante) [inline source/features/deep-reblame.tsx:88]: We should probably drop this. Buttons shouldn't disappear after being clicked. If the user missed the message they can click again and re-read it. It'll be handled by `mem` anyway

File: source/features/deep-reblame.tsx (hunk 9 above)
Comment (fregante) [reply source/features/deep-reblame.tsx:57]: You had this

https://github.com/sindresorhus/refined-github/blob/bc4938f6a7f79cd63379af31294e5f636de988c2/source/features/deep-blame.tsx#L53-54

**Pull request merged**